packages/flame/lib/src/components/component.dart

```diff
@@ -255,63 +277,86 @@ class Component {
-      // first parent.
-      component.onMount();
-      if (component.hasChildren) {
-        await component._reAddChildren();
-      }
-    }
-    children.addChild(component);
-  }
-
-  /// Adds multiple children.
+  /// The cost of this flexibility is that the component won't be added right
+  /// away. Instead, it will be placed into a queue, and then added later, after
+  /// it has finished loading, but no sooner than on the next game tick.
+  ///
+  /// When multiple children are scheduled to be added to the same parent, we
+  /// start loading all of them as soon as possible. Nevertheless, the children
+  /// will end up being added to the parent in exactly the same order as they
+  /// were originally scheduled by the user, regardless of how fast or slow
+  /// each of them loads.
+  ///
+  /// A component can be added to a parent which may not be mounted to the game
+  /// tree yet. In such case, the component will start loading immediately, but
+  /// its mounting will be delayed until such time when the parent becomes
+  /// mounted.
   ///
-  /// See [add] for details.
+  /// This method returns a future that completes when the component is done
+  /// loading, and mounting if the parent is currently mounted. However, this
+  /// future will not guarantee that the component will become "fully mounted":
+  /// it still needs to be added to the parent's children list, and that
+  /// operation will only be done on the next game tick.
+  ///
+  /// A component can only be added to one parent at a time. It is an error to
+  /// try to add it to multiple parents, or even to the same parent multiple
+  /// times. If you need to change the parent of a component, use the
+  /// [changeParent] method.
+  Future<void>? add(Component component) => component.addToParent(this);
+
+  /// A convenience method to [add] multiple children at once.
   Future<void> addAll(Iterable<Component> components) {
-    return Future.wait(components.map(add));
+    final futures = <Future<void>>[];
+    for (final component in components) {
+      final future = add(component);
+      if (future != null) {
+        futures.add(future);
+      }
+    }
+    return Future.wait(futures);
   }
 
-  /// The children are added again to the component set so that [prepare],
-  /// [onLoad] and [onMount] runs again. Used when a parent is changed
-  /// further up the tree.
-  Future<void> _reAddChildren() async {
-    if (_children != null) {
-      await Future.wait(_children!.map(add));
-      await Future.wait(_children!.addLater.map(add));
+  /// Adds this component to the provided [parent] (see [add] for details).
+  Future<void>? addToParent(Component parent) {
+    assert(
+      _parent == null,
+      '$this cannot be added to $parent because it already has a parent: '
+      '$_parent',
+    );
+    assert(
+      _state == LifecycleState.uninitialized ||
+          _state == LifecycleState.removed,
```

Comment:
pretty message for this assert? or this one that "should never happen AFAWK"

Reply:
This is just a sanity check on the internal state: I do not think it is possible to violate this assert in the current code. But, if in the future we need to refactor this again (say, adding more states), this assert may trigger.